verification/policy_file_processor.go

```diff
@@ -0,0 +1,348 @@
+	var filteredPolicies []*TrustPolicy
+	policies := pd.ListPolicies()
+
+	for _, policy := range policies {
+		if isMatchedScope(policy.RegistryScopes, scope) {
+			filteredPolicies = append(filteredPolicies, policy)
+		}
+	}
+
+	return filteredPolicies
+}
+
+// isMatchedScope returns true if the given scope matches the scope list.
+// A scope matches if either one holds true:
+//   1. scopes contain global scope.
+//   2. it's present in the scope list.
+func isMatchedScope(scopes []string, matchingScope string) bool {
+	if len(scopes) == 1 && scopes[0] == wildcard {
+		return true
+	}
+	return isPresent(matchingScope, scopes)
+}
+
+// UpdatePolicy updates existing policy with given new policy.
+// Notes: new policy only contains fields that need to be updated.
+func (pd *PolicyDocument) UpdatePolicy(policy *TrustPolicy) error {
+	existingPolicies := pd.getNameToPolicyMap()
+
+	if _, exist := existingPolicies[policy.Name]; !exist {
+		return ErrorPolicyNotExists{
+			Msg: fmt.Sprintf("%s not exists.", policy.Name),
+		}
+	}
+	mergedPolicy := mergePolicy(existingPolicies[policy.Name], policy)
+
+	return pd.updatePolicies([]*TrustPolicy{mergedPolicy})
+}
+
+// UpdatePolicies updates existing policies with given new policies.
+// Notes: new policies only contain fields that need to be updated.
+func (pd *PolicyDocument) UpdatePolicies(policies []*TrustPolicy) error {
+	existingPolicies := pd.getNameToPolicyMap()
+
+	for idx, policy := range policies {
+		if _, exist := existingPolicies[policy.Name]; !exist {
+			return ErrorPolicyNotExists{
+				Msg: fmt.Sprintf("%s not exists.", policy.Name),
+			}
+		}
+		policies[idx] = mergePolicy(existingPolicies[policy.Name], policy)
+	}
+
+	return pd.updatePolicies(policies)
+}
+
+// updatePolicies updates existing policies by replacing them with given
+// policies of the same name.
+func (pd *PolicyDocument) updatePolicies(policies []*TrustPolicy) error {
+	existingPolicies := pd.getNameToPolicyMap()
+
+	for _, policy := range policies {
+		if err := validatePolicy(policy); err != nil {
+			return err
+		}
+		existingPolicies[policy.Name] = policy
+	}
+
+	pd.overwritePolicies(existingPolicies)
+	return pd.save()
+}
+
+// DeletePolicies deletes specified policies from the PolicyDocument.
+func (pd *PolicyDocument) DeletePolicies(names []string) error {
+	existingPolicies := pd.getNameToPolicyMap()
+	uniqueNames := make(map[string]struct{})
+	for _, name := range names {
+		uniqueNames[name] = struct{}{}
+	}
+
+	for name, _ := range uniqueNames {
+		if _, exist := existingPolicies[name]; !exist {
+			return ErrorPolicyNotExists{
+				Msg: fmt.Sprintf("Policy %s does not exist", name),
+			}
+		}
+		delete(existingPolicies, name)
+	}
+	// return error if all policies are deleted.
+	if len(existingPolicies) == 0 {
+		return errors.New("illegal to delete all policies")
+	}
+
+	pd.overwritePolicies(existingPolicies)
+	return pd.save()
```

Comment:
It is better that the path to be write can be overwrite by notation-cli, because notation may support system level write in the future.

Reply:
good point, I'll add it.